shared/src/main/scala/mlscript/MLParser.scala

```diff
@@ -304,7 +305,186 @@ class MLParser(origin: Origin, indent: Int = 0, recordLocations: Bool = true) {
+        (tparams, tupBody)
+      // cases where the type is applied to type parameters or no parameters
+      // int
+      // int list
+      // (int, string) tup2
+      case (Seq(), S(t)) => (Set.empty[TypeName], t)
+      case (parts, S(t)) =>
+        val tparams = parts.flatMap(_._1).toSet
+        val args = parts.map(_._2).toList
+        (tparams, AppliedType(t, args))
+    } |
+      // type name or variable optionally applied to a type
+      ((tyName.map(L.apply) | adtTyParam.map(R.apply)) ~ tyName.rep()).map {
+        case (L(tname), Seq()) => (Set.empty[TypeName], tname)
+        case (R(tparam), Seq()) => (Set(tparam), tparam)
+        case (L(tparam), types) =>
+          val appType = types.foldLeft(tparam: Type) { case (appType, t) => AppliedType(t, appType :: Nil) }
+          (Set.empty[TypeName], appType)
+        case (R(tparam), types) =>
+          val appType = types.foldLeft(tparam: Type) { case (appType, t) => AppliedType(t, appType :: Nil) }
+          (Set(tparam), appType)
+      }
+
+  def adtTyFun[p: P]: P[(Set[TypeName], Type)] =
+    (adtTyExpr | ("(" ~/ adtTyExpr ~ ")")).rep(1, "->")
+      .map {
+        case Seq(t) => t
+        case parts =>
+          val tparams = parts.flatMap(_._1).toSet
+          val funBody = parts.init.map(_._2).foldRight(parts.last._2) {
+            case (arg, ret) => Function(toParamsTy(arg), ret)
+          }
+          (tparams, funBody)
+      }
+
+  /** Type alias body made of parts in a product type. each part can be a type
+    * or a function.
+    * Can return TypeName or a Tuple or a Function
+    */
+  def adtTyAlias[p: P]: P[(Set[TypeName], Type)] =
+    (adtTyFun | ("(" ~/ adtTyFun ~ ")")).rep(1, "*")
+      .map {
+        case Seq(t) => t
+        case parts =>
+          val tparams = parts.flatMap(_._1).toSet
+          val tupBody = Tuple(parts.map(part => (N, Field(N, part._2))).toList)
+          (tparams, tupBody)
+      }
+
+  /** data constructor body
+    * type 'a, 'b tup = Tup of ['a * 'b] => ('a, 'b)
+    * type 'a, 'b tup = Tup of ['a list * 'b] => (List['a], 'b)
+    * type 'a, 'b tup = Tup of ['a * 'b list] => ('a, List['b])
+    * type 'a, 'b tup = Tup of [('a * 'b) list] => (List[('a, 'b)])
+    */
+  def adtCtorBody[p: P]: P[(Set[TypeName], Record)] =
+    adtTyExpr.rep(1, "*")
+      .map {
+        case Seq((tparams, tbody)) =>
+          val rcdBody = Record(Var("_0") -> Field(N, tbody) :: Nil)
+          (tparams, rcdBody)
+        case parts =>
+          val tparams = parts.flatMap(_._1).toSet
+          val rcdBody = parts.zipWithIndex.map {
+            case (t, i) => Var(s"_$i") -> Field(N, t._2)
+          }.toList
+          (tparams, Record(rcdBody))
+      }
+
+  def ctorName[p: P]: P[TypeName] =
+    locate(P(uppercase ~~ (letter | digit | "_" | "'").repX).!.filter(!keywords(_)).map(TypeName))
+
+  /** data constructor declaration
+    * type 'a, 'b tup = [Tup of 'a * 'b]
```

Comment:
We already support type synonyms! There is no need to introduce an alternative syntax for them.

On the other hand, I am lost as to why you think `type ('a, 'b) tup = [Tup of 'a * 'b]` should be a valid type declaration. The `of` makes no sense in this position.

Reply:
The bracket [] shows the part of the statement that is parsed. I'll remove the syntax for type alias.